### ISSUE TITLE
Faster encoder and decoder model

### ISSUE THREAD

**Issue by zhangjinsong3** [MaybeShewill-CV/lanenet-lane-detection#200]
Thanks for your awesome work, And when will u have time to train a more faster model? such as mobilenet or nasnet.

**Comment by MaybeShewill-CV** [comment]
@zhangjinsong3 If I have spare time I will try it. I want to port the model into embedding device:)

**Comment by zacario-li** [comment]
@MaybeShewill-CV I ported the code with mobilenet on TX2, but the performance is bad. If I dropped the cluster steps, the performance is good, about 20fps.

**Comment by zhangjinsong3** [comment]
@zacario-li And how about the FPS of VGG based lanenet on TX2?

**Comment by MaybeShewill-CV** [comment]
@zacario-li Which procedure do you mean during cluster steps in specific and could you please paste some of your test result here and share with us:)

**Comment by yinhai86924** [comment]
> @MaybeShewill-CV I ported the code with mobilenet on TX2, but the performance is bad. If I dropped the cluster steps, the performance is good, about 20fps.
Why do you use mobileNet V2 speed only 20fps? Which layers do you extract for upsampling? Can you share your encoder code using mobilenetV2?

**Comment by yinhai86924** [comment]
> @MaybeShewill-CV I ported the code with mobilenet on TX2, but the performance is bad. If I dropped the cluster steps, the performance is good, about 20fps.



> @MaybeShewill-CV I ported the code with mobilenet on TX2, but the performance is bad. If I dropped the cluster steps, the performance is good, about 20fps.

Why do you use mobileNet V2 speed only 20fps? Which layers do you extract for upsampling? Can you share your encoder code using mobilenetV2?

**Comment by zacario-li** [comment]
@zhangjinsong3 if use vgg, it will OOM

**Comment by zacario-li** [comment]
@yinhai86924 
mobilenetv2 use these layers below for upsampling:
`decode_layer_list=['layer_19',
                                 'layer_14',
                                  'layer_7']`

**Comment by zhangjinsong3** [comment]
@zacario-li That is strange. It only occupy 2.5GB GPU memory on my PC, with a speed of 0.037s per frame. My GPU is 1050Ti.

**Comment by yinhai86924** [comment]
> @yinhai86924
> mobilenetv2 use these layers below for upsampling:
> `decode_layer_list=['layer_19', 'layer_14', 'layer_7']`

Thank you very much! I hope you can share the code using mobileNetV2 as the enconder phase on GitHub? Let's talk about it.

**Comment by yinhai86924** [comment]
> @zacario-li That is strange. It only occupy 2.5GB GPU memory on my PC, with a speed of 0.037s per frame. My GPU is 1050Ti.

1060ti only 4.97G，0.05s per frame

**Comment by MaybeShewill-CV** [comment]
@zacario-li Which procedure do you mean during cluster steps in specific and could you please paste some of your test result here and share with us:)

**Comment by zacario-li** [comment]
@MaybeShewill-CV cluster.get_lane_mask() cost about 0.1s depend on binary_seg_ret's value. If binary seg has lots of lane mask, that will cost more time than less lane mask.

**Comment by zacario-li** [comment]
@yinhai86924 
`
from encoder_decoder_model import cnn_basenet, mobilenet_v2

class MobilenetEncoder(cnn_basenet.CNNBaseModel):
    def __init__(self,phase):
        super(MobilenetEncoder,self).__init__()
        self._train_phase = tf.constant('train',dtype=tf.string)
        self._test_phase = tf.constant('test',dtype=tf.string)
        self._phase = phase
        self._is_training = self.__init_phase__()

    def __init_phase__(self):
        return tf.equal(self._phase,self._train_phase)

    def encode(self,input_tensor,name):
        ret = OrderedDict()
        with tf.variable_scope(name):
            net , end_points = mobilenet_v2.mobilenet_base(input_tensor)
            #prepare output feature maps
            for i in range(1,20):
                ret['layer_{}'.format(i)] = dict()
                ret['layer_{}'.format(i)]['data'] = end_points['layer_{}'.format(i)]
                ret['layer_{}'.format(i)]['shape'] = end_points['layer_{}'.format(i)].get_shape().as_list()
        return ret
`

**Comment by zacario-li** [comment]
@zhangjinsong3 
If I only run lanenet model, and don't do the lane cluster, the speed is 0.019s per frame on P40

**Comment by yinhai86924** [comment]
> @yinhai86924
> `
> from encoder_decoder_model import cnn_basenet, mobilenet_v2
> 
> class MobilenetEncoder(cnn_basenet.CNNBaseModel):
> def **init**(self,phase):
> super(MobilenetEncoder,self).**init**()
> self._train_phase = tf.constant('train',dtype=tf.string)
> self._test_phase = tf.constant('test',dtype=tf.string)
> self._phase = phase
> self._is_training = self.**init_phase**()
> 
> ```
> def __init_phase__(self):
>     return tf.equal(self._phase,self._train_phase)
> 
> def encode(self,input_tensor,name):
>     ret = OrderedDict()
>     with tf.variable_scope(name):
>         net , end_points = mobilenet_v2.mobilenet_base(input_tensor)
>         #prepare output feature maps
>         for i in range(1,20):
>             ret['layer_{}'.format(i)] = dict()
>             ret['layer_{}'.format(i)]['data'] = end_points['layer_{}'.format(i)]
>             ret['layer_{}'.format(i)]['shape'] = end_points['layer_{}'.format(i)].get_shape().as_list()
>     return ret
> ```
> `

I really thank you !!! how to define the mobilenet_base? can you share mobilenet_base code?

**Comment by zacario-li** [comment]
> > @yinhai86924
> > `
> > from encoder_decoder_model import cnn_basenet, mobilenet_v2
> > class MobilenetEncoder(cnn_basenet.CNNBaseModel):
> > def **init**(self,phase):
> > super(MobilenetEncoder,self).**init**()
> > self._train_phase = tf.constant('train',dtype=tf.string)
> > self._test_phase = tf.constant('test',dtype=tf.string)
> > self._phase = phase
> > self._is_training = self.**init_phase**()
> > ```
> > def __init_phase__(self):
> >     return tf.equal(self._phase,self._train_phase)
> > 
> > def encode(self,input_tensor,name):
> >     ret = OrderedDict()
> >     with tf.variable_scope(name):
> >         net , end_points = mobilenet_v2.mobilenet_base(input_tensor)
> >         #prepare output feature maps
> >         for i in range(1,20):
> >             ret['layer_{}'.format(i)] = dict()
> >             ret['layer_{}'.format(i)]['data'] = end_points['layer_{}'.format(i)]
> >             ret['layer_{}'.format(i)]['shape'] = end_points['layer_{}'.format(i)].get_shape().as_list()
> >     return ret
> > ```
> > `
> 
> I really thank you !!! how to define the mobilenet_base? can you share mobilenet_base code?

It copied from the tensorflow's official github, link is [here](https://github.com/tensorflow/models/tree/master/research/slim/nets/mobilenet)

**Comment by yinhai86924** [comment]
> > > @yinhai86924
> > > `
> > > from encoder_decoder_model import cnn_basenet, mobilenet_v2
> > > class MobilenetEncoder(cnn_basenet.CNNBaseModel):
> > > def **init**(self,phase):
> > > super(MobilenetEncoder,self).**init**()
> > > self._train_phase = tf.constant('train',dtype=tf.string)
> > > self._test_phase = tf.constant('test',dtype=tf.string)
> > > self._phase = phase
> > > self._is_training = self.**init_phase**()
> > > ```
> > > def __init_phase__(self):
> > >     return tf.equal(self._phase,self._train_phase)
> > > 
> > > def encode(self,input_tensor,name):
> > >     ret = OrderedDict()
> > >     with tf.variable_scope(name):
> > >         net , end_points = mobilenet_v2.mobilenet_base(input_tensor)
> > >         #prepare output feature maps
> > >         for i in range(1,20):
> > >             ret['layer_{}'.format(i)] = dict()
> > >             ret['layer_{}'.format(i)]['data'] = end_points['layer_{}'.format(i)]
> > >             ret['layer_{}'.format(i)]['shape'] = end_points['layer_{}'.format(i)].get_shape().as_list()
> > >     return ret
> > > ```
> > > `
> > 
> > 
> > I really thank you !!! how to define the mobilenet_base? can you share mobilenet_base code?
> 
> It copied from the tensorflow's official github, link is [here](https://github.com/tensorflow/models/tree/master/research/slim/nets/mobilenet)

Official MobileNet_v2 is complex, which modules are you copied？
![QQ截图20190412164811](https://user-images.githubusercontent.com/45553479/56024826-675f4680-5d43-11e9-9384-45f138e643ab.png)

**Comment by zacario-li** [comment]
> > > > @yinhai86924
> > > > `
> > > > from encoder_decoder_model import cnn_basenet, mobilenet_v2
> > > > class MobilenetEncoder(cnn_basenet.CNNBaseModel):
> > > > def **init**(self,phase):
> > > > super(MobilenetEncoder,self).**init**()
> > > > self._train_phase = tf.constant('train',dtype=tf.string)
> > > > self._test_phase = tf.constant('test',dtype=tf.string)
> > > > self._phase = phase
> > > > self._is_training = self.**init_phase**()
> > > > ```
> > > > def __init_phase__(self):
> > > >     return tf.equal(self._phase,self._train_phase)
> > > > 
> > > > def encode(self,input_tensor,name):
> > > >     ret = OrderedDict()
> > > >     with tf.variable_scope(name):
> > > >         net , end_points = mobilenet_v2.mobilenet_base(input_tensor)
> > > >         #prepare output feature maps
> > > >         for i in range(1,20):
> > > >             ret['layer_{}'.format(i)] = dict()
> > > >             ret['layer_{}'.format(i)]['data'] = end_points['layer_{}'.format(i)]
> > > >             ret['layer_{}'.format(i)]['shape'] = end_points['layer_{}'.format(i)].get_shape().as_list()
> > > >     return ret
> > > > ```
> > > > `
> > > 
> > > 
> > > I really thank you !!! how to define the mobilenet_base? can you share mobilenet_base code?
> > 
> > 
> > It copied from the tensorflow's official github, link is [here](https://github.com/tensorflow/models/tree/master/research/slim/nets/mobilenet)
> 
> Official MobileNet_v2 is complex, which modules are you copied？
> ![QQ截图20190412164811](https://user-images.githubusercontent.com/45553479/56024826-675f4680-5d43-11e9-9384-45f138e643ab.png)

conv_blocks.py, mobilenet.py, mobilenet_v2.py

**Comment by yinhai86924** [comment]
> @yinhai86924
> mobilenetv2 use these layers below for upsampling:
> `decode_layer_list=['layer_19', 'layer_14', 'layer_7']`

I don't know how to modify this ： Load mobilenet pre-training weights  can you share this code？
# 加载预训练参数
        if net_flag == 'vgg' and weights_path is None:
            pretrained_weights = np.load(
                './data/vgg16.npy',
                encoding='latin1').item()

            for vv in tf.trainable_variables():
                weights_key = vv.name.split('/')[-3]
                try:
                    weights = pretrained_weights[weights_key][0]
                    _op = tf.assign(vv, weights)
                    sess.run(_op)
                except Exception as e:
                    continue

**Comment by jackofall96** [comment]
> > @yinhai86924
> > mobilenetv2 use these layers below for upsampling:
> > `decode_layer_list=['layer_19', 'layer_14', 'layer_7']`
> 
> how to Load mobilenet_v2 pre-training weights? can you share this code? Pre-training weights is mobilenet_v2_1.0_224.ckpt? I try it but not correct , said“” The passed save_path is not a valid checkpoint: E:\new-Lanenet\lanenet-lane-detection-master\mobilenet_v2_1.0_224.ckpt”
> Thank you for your help!
> I don't know how to modify this ：
> 
> # 加载预训练参数
> ```
>     if net_flag == 'vgg' and weights_path is None:
>         pretrained_weights = np.load(
>             './data/vgg16.npy',
>             encoding='latin1').item()
> 
>         for vv in tf.trainable_variables():
>             weights_key = vv.name.split('/')[-3]
>             try:
>                 weights = pretrained_weights[weights_key][0]
>                 _op = tf.assign(vv, weights)
>                 sess.run(_op)
>             except Exception as e:
>                 continue
> ```

I'm facing the same issue, were you able to fix it?

**Comment by jackofall96** [comment]
> @yinhai86924
> mobilenetv2 use these layers below for upsampling:
> `decode_layer_list=['layer_19', 'layer_14', 'layer_7']`

Any reason why you particularly used these 3 layers?

**Comment by yinhai86924** [comment]
> I don't know how to modify this ：
> 加载预训练参数
>     if net_flag == 'vgg' and weights_path is None:
>         pretrained_weights = np.load(
>             './data/vgg16.npy',
>             encoding='latin1').item()
> 
>         for vv in tf.trainable_variables():
>             weights_key = vv.name.split('/')[-3]
>             try:
>                 weights = pretrained_weights[weights_key][0]
>                 _op = tf.assign(vv, weights)
>                 sess.run(_op)
>             except Exception as e:
>                 continue
> 
> 
> @yinhai86924 I'm facing the same issue, were you able to fix it?
I do not use the official code of mobilenet_v2,  becouse official code is based on TF-slim, therefore, I write the mobilenet_v2 by myself.

**Comment by jackofall96** [comment]
> > I do not use the official code of mobilenet_v2,  becouse official code is based on TF-slim, therefore, I write the mobilenet_v2 by myself.
But how do you assign the weights to their nodes?

**Comment by yinhai86924** [comment]
> I do not use the official code of mobilenet_v2,  becouse official code is based on TF-slim, therefore, I write the mobilenet_v2 by myself.
> 
> 
> @yinhai86924
> But how do you assign the weights to their nodes?
> And are you loading the weights from mobilenet_v2_1.0_224.ckpt?
> When I loaded the weights from the frozen graph of mobilenet_v2, all the weights that had to be updated were all named 'Const_{number}'. I could not tell which weights belonged to which trainable variables.
> If anyone managed to do this, please do help out.
Do you use official mobilenet_v2 code？

**Comment by yinhai86924** [comment]
> I do not use the official code of mobilenet_v2,  becouse official code is based on TF-slim, therefore, I write the mobilenet_v2 by myself.
> @yinhai86924
> But how do you assign the weights to their nodes?
> And are you loading the weights from mobilenet_v2_1.0_224.ckpt?
> When I loaded the weights from the frozen graph of mobilenet_v2, all the weights that had to be updated were all named 'Const_{number}'. I could not tell which weights belonged to which trainable variables.
> If anyone managed to do this, please do help out.
> Do you use official mobilenet_v2 code？
> 
> 
> @yinhai86924
> Yes, I do. The one using TF_slim, as suggested above.
> Were you able to train using the code you wrote yourself? How did you assign the weights there?

You can refer to the structure of mobilenetv2 and write it yourself to distribute weights. I use official code , but, can not load weights

**Comment by jingbanghu** [comment]
> @yinhai86924
> `
> from encoder_decoder_model import cnn_basenet, mobilenet_v2
> 
> class MobilenetEncoder(cnn_basenet.CNNBaseModel):
> def **init**(self,phase):
> super(MobilenetEncoder,self).**init**()
> self._train_phase = tf.constant('train',dtype=tf.string)
> self._test_phase = tf.constant('test',dtype=tf.string)
> self._phase = phase
> self._is_training = self.**init_phase**()
> 
> ```
> def __init_phase__(self):
>     return tf.equal(self._phase,self._train_phase)
> 
> def encode(self,input_tensor,name):
>     ret = OrderedDict()
>     with tf.variable_scope(name):
>         net , end_points = mobilenet_v2.mobilenet_base(input_tensor)
>         #prepare output feature maps
>         for i in range(1,20):
>             ret['layer_{}'.format(i)] = dict()
>             ret['layer_{}'.format(i)]['data'] = end_points['layer_{}'.format(i)]
>             ret['layer_{}'.format(i)]['shape'] = end_points['layer_{}'.format(i)].get_shape().as_list()
>     return ret
> ```
> 
> `

how to do in decode?can u show the code? thx!
